libgalois/src/PropertyGraph.cpp

```diff
@@ -145,6 +146,182 @@ MakePropertyGraph(
       std::move(rdg_file), std::move(rdg_result.value()));
 }
 
+/// Assumes all boolean or uint8 properties are types
+katana::Result<std::shared_ptr<arrow::NumericArray<arrow::UInt8Type>>>
+GetTypeIDsFromProperties(
+    const std::shared_ptr<arrow::Table>& properties,
+    katana::PropertyGraph::TypeToTypeIDsMap* type_name_to_ids,
+    katana::PropertyGraph::TypeIDToTypesMap* type_id_to_names) {
+  // throw an error if each column/property has more than 1 chunk
+  for (int i = 0, n = properties->num_columns(); i < n; i++) {
+    std::shared_ptr<arrow::ChunkedArray> property = properties->column(i);
+    if (property->num_chunks() != 1) {
+      return KATANA_ERROR(
+          katana::ErrorCode::NotImplemented,
+          "execution failed because property {} has {} chunks (1 chunk "
+          "expected)",
+          properties->schema()->field(i)->name(), property->num_chunks());
+    }
+  }
+
+  // collect the list of types
+  std::vector<int> type_field_indices;
+  const std::shared_ptr<arrow::Schema>& schema = properties->schema();
+  KATANA_LOG_DEBUG_ASSERT(schema->num_fields() == properties->num_columns());
+  for (int i = 0, n = schema->num_fields(); i < n; i++) {
+    const std::shared_ptr<arrow::Field>& current_field = schema->field(i);
+
+    // a bool or uint8 property is (always) considered a type
+    // TODO(roshan) make this customizable by the user
+    if (current_field->type()->Equals(arrow::boolean()) ||
+        current_field->type()->Equals(arrow::uint8())) {
+      type_field_indices.push_back(i);
+    }
+  }
+
+  // assign a new ID to each type
+  // NB: cannot use unordered_map without defining a hash function for vectors;
+  // performance is not affected here because the map is very small (<=256)
+  std::map<std::vector<int>, katana::PropertyGraph::TypeID>
+      type_field_indices_to_id;
+  for (int i : type_field_indices) {
+    katana::PropertyGraph::TypeID new_type_id = type_id_to_names->size();
+    std::vector<int> field_indices = {i};
+    type_field_indices_to_id.emplace(
+        std::make_pair(field_indices, new_type_id));
+
+    const std::shared_ptr<arrow::Field>& current_field = schema->field(i);
+    const std::string& field_name = current_field->name();
+    KATANA_LOG_DEBUG_ASSERT(
+        type_name_to_ids->find(field_name) == type_name_to_ids->end());
+
+    std::unordered_set<katana::PropertyGraph::TypeID> type_ids{new_type_id};
+    type_name_to_ids->emplace(std::make_pair(field_name, type_ids));
+    type_id_to_names->push_back({field_name});
+  }
+
+  // collect the list of unique combination of types
+  // NB: cannot use unordered_set without defining a hash function for vectors;
+  // performance is not affected here because the set is very small (<=256)
+  std::set<std::vector<int>> type_combinations;
+  katana::PerThreadStorage<std::set<std::vector<int>>> type_combinations_pts;
```

Comment:
gstl::Set and gstl::Vector as these data structures (set & vector) will allocate memory in parallel section when populating them.